ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java

```diff
@@ -150,17 +159,46 @@ public static String addContextToPrompt(Map<String, String> parameters, String p
         return prompt;
     }
 
+    public static List<String> MODEL_RESPONSE_PATTERNS = List
+        .of(
+            "\\{\\s*\"thought\":.*?\\s*,\\s*\"action\":.*?\\s*,\\s*\"action_input\":.*?\\}",
+            "\\{\\s*\"thought\"\\s*:\\s*\".*?\"\\s*,\\s*\"action\"\\s*:\\s*\".*?\"\\s*,\\s*\"action_input\"\\s*:\\s*\".*?\"\\s*}",
+            "\\{\\s*\"thought\"\\s*:\\s*\".*?\"\\s*,\\s*\"final_answer\"\\s*:\\s*\".*?\"\\s*}"
+        );
+
     public static String extractModelResponseJson(String text) {
-        Pattern pattern = Pattern.compile("```json\\s*([\\s\\S]+?)\\s*```");
-        Matcher matcher = pattern.matcher(text);
+        return extractModelResponseJson(text, null);
+    }
 
-        if (matcher.find()) {
-            return matcher.group(1);
+    public static String extractModelResponseJson(String text, List<String> llmResponsePatterns) {
+        Pattern pattern1 = Pattern.compile("```json\\s*([\\s\\S]+?)\\s*```");
+        Matcher matcher1 = pattern1.matcher(text);
+
+        if (matcher1.find()) {
+            return matcher1.group(1);
         } else {
+            String matchedPart = findMatchedPart(text, MODEL_RESPONSE_PATTERNS);
+            if (matchedPart == null && llmResponsePatterns != null) {
```

Comment:
add a comment like: `// If no match is found, try additional patterns if provided`